docs/proposals/flexvolume-api.md

```diff
@@ -0,0 +1,177 @@
+* Add support for dynamic provisioning and centralized attach-detach controller and
+* The biggest advantage of flex volume is it’s simplicity. It is just an adapter around the internal api. We will keep it simple in the new design and try to follow the same adapter pattern.
+
+### Non Goals:
+* This proposal does not the address the poor installation model of existing model. It is covered in another PR #32550.
+
+## Terminology:
+**Plugin**: refers to flex volume plugin.
+
+**Plugin Driver API**: API calls defined by flex volume plugin.
+
+**Plugin Driver**: Out-of-tree driver/plugin which implements plugin driver api.
+
+**In-tree plugin**: Plugins which are included in Kubernetes project/source tree.
+
+## API:
+Existing flex volume plugin driver api is in alpha phase and are subject to change between Kubernetes versions. This section proposes stable api for plugin driver writers to work with.
+
+### Proposal:
+Proposes fewer APIs with stable request & response API objects.
+
+###### Pros:
+* Has standard request & response API objects.
+* Supports versioning.
+* Uses gRPC.
+* Fewer calls.
+* Simple and clean interface.
+
+###### Cons:
+* New API objects.
+
+#### APIs
+
+```protobuf
+
+// Capabilities of the driver
+message DriverCapabilities {
+	string supported_read_write_modes = 1;
+	enum AttachmentPolicy {
+	    NONE = 0;  // Driver does not support attach/detach.
+	    LOCAL = 1; // Driver supports local volume attachments from kubelet.
+	    REMOTE = 2; // Driver supports remote volume attachments to a node from controller-manager.
+	}
+	AttachmentPolicy supported_attachment_policy = 2;
+	bool supports_dynamic_provisioning = 3;
+	bool supports_custom_mount = 4;
+	bool supports_selinux = 5;
+	bool supports_ownership_management = 6;
+	bool supports_metrics = 7;
+}
+
+message ProbeDriverRequest {}
+
+message ProbeDriverResponse {
+    DriverCapabilities capabilities = 1;
+    repeated string options = 2;
+}
+
+service FlexVolumeService {
+    // Probe the driver capabilities.
+    rpc ProbeDriver(ProbeDriverRequest) returns (ProbeDriverResponse) {}
+    // Create a volume.
+    rpc Create(CreateRequest) returns (CreateResponse) {}
+    // Delete a volume.
+    rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+    // Attach the volume.
+    rpc Attach(AttachRequest) returns (AttachResponse) {}
+    // Detach the volume.
+    rpc Detach(DetachRequest) returns (DetachResponse) {}
+    // Mount the volume.
+    rpc Mount(MountRequest) returns (MountResponse) {}
+    // unmount the volume.
+    rpc Unmount(UnmountRequest) returns (UnmountResponse) {}
+    // Query the metrics.
+    rpc GetMetrics(MetricsRequest) returns (MetricsResponse) {}
+}
+
+// Volume source.
+message Spec {
+    string name = 1; // Name of the volume.
+    string fstype = 2; // Requested file system type.
+    string read_write_mode = 3;
+    map<string, string> secrets = 4; // Secrets required for the driver to talk to its own controller.
```

Comment:
Very good point. I should change the flex volume source to reflect it. The plan is to pass the secret value to the plugins though.